lib/rack/utils.rb

```diff
@@ -414,6 +414,12 @@ def initialize_copy(other)
         @names = other.names.dup
       end
 
+      # on clear, we need to clear @names hash
+      def clear
+        super
+        @names = {}
```

Comment:
We can clear the hash without allocating a new one: [`Hash#clear`](http://ruby-doc.org/core-2.2.0/Hash.html#method-i-clear)